### PR TITLE
[SPARK-18871][SQL][TESTS] New test cases for IN/NOT IN subquery 3rd batch

### PR DESCRIPTION
**Pull request by kevinyu98**
## What changes were proposed in this pull request?

This is 3ird batch of test case for IN/NOT IN subquery. In this PR, it has these test files:

`in-having.sql`
`in-joins.sql`
`in-multiple-columns.sql`

These are the queries and results from running on DB2.
[in-having DB2 version](https://github.com/apache/spark/files/772668/in-having.sql.db2.txt)
[output of in-having](https://github.com/apache/spark/files/772670/in-having.sql.db2.out.txt)
[in-joins DB2 version](https://github.com/apache/spark/files/772672/in-joins.sql.db2.txt)
[output of in-joins](https://github.com/apache/spark/files/772673/in-joins.sql.db2.out.txt)
[in-multiple-columns DB2 version](https://github.com/apache/spark/files/772678/in-multiple-columns.sql.db2.txt)
[output of in-multiple-columns](https://github.com/apache/spark/files/772680/in-multiple-columns.sql.db2.out.txt)

## How was this patch tested?
This pr is adding new test cases. We compare the result from spark with the result from another RDBMS(We used DB2 LUW). If the results are the same, we assume the result is correct.

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-having.sql.out [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+WHERE  t1b > 0
+GROUP  BY t1a
+HAVING t1a NOT IN (SELECT t2a
+                   FROM   t2
+                   WHERE  t2b > 3)
+-- !query 11 schema
+struct<t1a:string,max(t1b):smallint>
+-- !query 11 output
+val1a	16
+val1d	10
```
Comment (nsyca) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-having.sql.out]: The results match with the ones from DB2.

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-joins.sql.out [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+and t1a = t2a
+Group By t1a, t1b, t1c, t2a, t2b, t2c
+HAVING t2c IS NOT NULL
+ORDER By t2b DESC nulls last
+-- !query 13 schema
+struct<t1a:string,t1b:smallint,t1c:int,count(DISTINCT t2a):bigint,t2b:smallint,t2c:int>
+-- !query 13 output
+val1b	8	16	1	10	12
+val1b	8	16	1	8	16
+val1b	8	16	1	NULL	16
```
Comment (nsyca) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-joins.sql.out]: The results match with the ones from DB2.

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-multiple-columns.sql.out [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                           FROM       cte1
+                           JOIN       cte1 cte2
+                           on         cte1.t1b = cte2.t1b) s
+-- !query 7 schema
+struct<t1a:string,t1b:smallint,t1a:string,t1b:smallint>
+-- !query 7 output
+val1b	8	val1b	8
+val1b	8	val1c	8
+val1c	8	val1b	8
+val1c	8	val1c	8
```
Comment (nsyca) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-multiple-columns.sql.out]: The results match with the ones from DB2.

**Comment by gatorsmile** [comment]
OK to test

**Comment by gatorsmile** [comment]
test this please

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-multiple-columns.sql.out [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                            FROM   t2
+                            WHERE  t2a = t1a
+                            ORDER  BY t2a)
+AND t1a = 'val1a'
+-- !query 3 schema
+struct<t1a:string,t1b:smallint,t1h:timestamp>
+-- !query 3 output
+val1a	16	2014-06-04 01:02:00.001
+val1a	16	2014-07-04 01:01:00
+val1a	6	2014-04-04 01:00:00
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-multiple-columns.sql.out]: This does not match the DB2 output.

File: sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-multiple-columns.sql [hunk 5]
```diff
@@ -0,0 +1,127 @@
+-- A test suite for multiple columns in predicate in parent side, subquery, and both predicate subquery
+-- It includes correlated cases.
+
+create temporary view t1 as select * from values
+  ("val1a", 6S, 8, 10L, float(15.0), 20D, 20E2, timestamp '2014-04-04 01:00:00.000', date '2014-04-04'),
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-multiple-columns.sql]: The values of timestamp do not match what you insert to DB2

Comment (kevinyu98) [inline sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-multiple-columns.sql]: Hello Sean: Thank for catching this. The db2 data is 
insert into t1 values ('val1a', 6, 8, 10, float(15.0), 20, 20E2, timestamp('2014-04-04 00:00:00.000'), date('2014-04-04'))..
I have verified that is the only difference in the data. I have updated the db2 data, and run against the db2 for all the test files, and verify the result are the same. I upload the updated db2 files.

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-multiple-columns.sql.out (hunk 4 above)
Comment (kevinyu98) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-multiple-columns.sql.out]: This is the same issue for the input data. After I updated the data for db2 test cases, the result is same now. I have upload the db2 files

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
test this please

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

File: sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-multiple-columns.sql (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-multiple-columns.sql]: Thanks!

**Comment by robbinspg** [comment]
@kevinyu98 Several of the new tests fail on Big Endian platforms. It appears that rows are returned in a slightly different order but are still a correct output from the query. For example in-joins query 4:


-- !query 4
SELECT    Count(DISTINCT(t1a)),
          t1b,
          t3a,
          t3b,
          t3c
FROM      t1 natural left JOIN t3
WHERE     t1a IN
          (
                 SELECT t2a
                 FROM   t2
                 WHERE t1d = t2d)
AND       t1b > t3b
GROUP BY  t1a,
          t1b,
          t3a,
          t3b,
          t3c
ORDER BY  t1a DESC


on Little Endian returns
1	10	val3b	8	NULL
1	10	val1b	8	16
1	10	val3a	6	12
1	8	val3a	6	12
1	8	val3a	6	12

wheras on big endian returns:
1	10	val3a	6	12
1	10	val3b	8	NULL
1	10	val1b	8	16
1	8	val3a	6	12
1	8	val3a	6	12

I believe GROUP BY does not define any ordering so both of these outputs are valid  for the query  as the ORDER BY is only on t1a but obviously the big endian output does not match your expected output so fails.

I'm trying to determine why the execution on big endian returns the rows in a different order.

**Comment by kevinyu98** [comment]
Hello Pete: Thanks for running the test case. Can you send the failing test case file to me? Also I can provide new test files with the output files, can you help test on your platforms? thanks.

**Comment by gatorsmile** [comment]
To make the results consistent between big endian and small endian, we can improve the queries with the extra order by clauses.

@robbinspg Which queries failed? @kevinyu98 Can you collect the failed cases and submit another PR for resolving the issues? Thanks!

**Comment by kevinyu98** [comment]
@gatorsmile sure, I will do that. Thanks.

**Comment by robbinspg** [comment]
OK I'll raise a separate Jira, document the differences and submit a PR

**Pull request closed without merging**